magma/lib/magma/migration.rb

```diff
@@ -152,7 +172,13 @@ def column_entry name, type
     end
 
     def remove_column_entry name
-      "drop_column :#{name}"
+      now = DateTime.now.to_time.to_i
+      new_name = "#{name}_#{now}_backup"
+
+      [
+        "drop_foreign_constraint_if_exists :#{name}",
```

Comment:
Does dropping the foreign constraint affect the existing data in that column at all?

Separately, is there some other entry that has to get modified in the `attributes` table, in this case? I think there might be a reference from the foreign table attribute ... i.e. if

`sc_seq` --> `sc_seq_pool`

And you drop the `sc_seq_pool` attribute on `sc_seq`, not only do you have to change the `sc_seq_pool` attribute where `model_name=sc_seq`, but will you also have to remove the `sc_seq` collection attribute where `model_name=sc_seq_pool`? Hopefully that makes sense...

Reply:
Yeah I'm concerned we'll have to do this in two parts, the easy part is removing data attributes, the hard part is unlinking models, which has all sorts of issues if we are touching parent attributes (like, how to maintain a connected graph while changing fundamental relations) - I'd say `remove_attribute` should only deal with the former.

Reply:
> Does dropping the foreign constraint affect the existing data in that column at all?

Nope, in fact, removing the constraint is what allows us to preserve the existing data better.  Basically, the data can remain always preserved and not subject the model to update failures from an invisible attribute.  If we don't remove the constraint, we are forced to  add data to the 'removed' relation on every new record (because nil or 0 won't work if it's not a real relationship).

Reply:
> Yeah I'm concerned we'll have to do this in two parts, the easy part is removing data attributes, the hard part is unlinking models, which has all sorts of issues if we are touching parent attributes (like, how to maintain a connected graph while changing fundamental relations) - I'd say remove_attribute should only deal with the former.

What is the goal of remove attribute?  I'd say, the focus of remove attribute is primarily to clean and refactor models when we need to make major changes.  In that case, I really feel handing it clear, useful power is the correct approach here.  I think we could spend a lot of time over engineering tools that prevent 'bad cases' but also don't solve 'necessary cases'.  

Here's my thought: removing a parent attribute and unlinking a model has no catostrophic effect other than to temporarily unlink the model until a new parent attribute is applied.  No data loss, no system downtime, just a temporarily inconvenient state that can be repaired.  BUT, it might be the obvious natural path in a multi step process to making a refactor of some other sort.

I think we should allow that step, because the alternative feels like additional work that will result in an unintuitive tool.  Eg: autocomplete trying to fight me about what I'm trying to type.  Sure, it might prevent some mistakes, that were never a big deal anyways, but it also definitely causes me "flight blindness" when I expected typing A -> "A", but I get A -> "A....<bs><bs><bs>Furniture"